packages/scratch-vm/src/extensions/scratch3_spotify/index.ts

```diff
@@ -0,0 +1,492 @@
+      playBeatAndWait: playBeatAndWait,*/
+            trackData: getTrackData,
+            /*currentBeat: getCurrentBeat,
+      everyBeat: everyBeat,
+      everyBar:everyBar,
+      stopMusic: stopMusic*/
+        };
+    }
+}
+
+// FEEDBACK is it redundant to have to define all of these blocks then make the function for them?
+// Example 4
+//type WithOptionsBlock = Blocks["exampleHat"];
+/*const pickFromOptions = (): Block<WithOptionsBlock> => ({
+  type: BlockType.Hat,
+  arg: { type: ArgumentType.Boolean, options: [{ text: 'Yes', value: true }, { text: 'No', value: false }] },
+  text: (argument1) => `Should the below block execute? ${argument1}`,
+  operation: function (argument1) {
+    return argument1;
+  }
+});*/
+
+type AccessToken = {
+    expirationTime: number;
+    value: string;
+};
+
+const currentTimeSec = (): number => {
+    return new Date().getTime() / 1000;
+};
+
+const getAccessToken = (): Promise<AccessToken> => {
+    return new Promise((resolve, reject) => {
+        nets(
+            {
+                // We're piggybacking off of Eric's (?) account and that's bad
+                // TODO make our own way to get a spotify token
+                url: "https://u61j2fb017.execute-api.us-east-1.amazonaws.com/prod/get-spotify-token",
+                timeout: SERVER_TIMEOUT,
+            },
+            (err, res, body) => {
+                if (err) {
+                    console.error("Spotify token error: " + err);
+                    // TODO make sure I did this right (changed from "return reject()")
+                    return reject();
+                }
+
+                if (res.statusCode !== 200) {
+                    console.error("Spotify token error: " + res.statusCode);
+                    return reject();
+                }
+
+                // success
+                let token: AccessToken = {
+                    expirationTime: currentTimeSec() + 3600,
+                    value: JSON.parse(body).token,
+                };
+                resolve(token);
+            }
+        );
+    });
+};
+
+const refreshAccessTokenIfNeeded = (
+    token: AccessToken
+): Promise<AccessToken> => {
+    return new Promise((resolve, reject) => {
+        if (currentTimeSec() > token.expirationTime) {
+            getAccessToken().then((newToken) => {
+                token = newToken;
+                console.log("token expired, got a new one");
+                resolve(token);
+            });
+        } else {
+            resolve(token);
+        }
+    });
+};
+
+type SearchBlock = Blocks["searchAndPlay"];
+const searchAndPlay = (Spotify): Block<SearchBlock> => ({
+    type: BlockType.Command,
+    arg: { type: ArgumentType.String, defaultValue: "tacos" },
+    text: (searchQuery) => `play music like ${searchQuery}`,
+    operation: function (searchQuery) {
+        console.log("play music like " + searchQuery);
+
+        // FEEDBACK how do I pass variables from the Spotify class to this function
```

Comment:
Seems like you got it, right? As you're pulling the `spotifyToken` off of your `Spotify` instance.

If not, let me know!

Reply:
Yup and I will clean up my comments next round

Reply:
Great point! Will definitely add. 

Do you think that's something best suited to go in the README? Or in the comments of the 'template' file you receive after running `npm run new:extension`

Reply:
Additional idea! I'll also add an FAQ section to the readme and add this item